### PR TITLE
Divrem speedup and tests

### PR DESCRIPTION
**Pull request by r7rohan**
[#22219](https://github.com/JuliaLang/julia/issues/22219#issue-233492600)
- Replaced rem call with `a-d*b`. Reduced div calls from 2-->1 (code in: `base\div.jl`)
Int-Int : Works for all rounding modes and speeds up by a factor of 1.6x
Floats - Abstract Floats, Abstract Irrational : Works but there is marginal error in Float64, BigFloat. (Not added)
Rationals - `a-d*b` is more expensive than normal div and rem calls. (Not added)

- Also added tests for divrem in rounding modes in `test\numbers.jl`

Passes all the tests for div, divrem, rem ` SUCCESS`
Also speeds divrem by a factor of 2
```
function divrem!(a, b, c, d)
    for i = 1:100_000
        c[i], d[i] = divrem(a[i], b[i])
    end
end

function divrem2!(a, b, c, d)
    for i = 1:100_000
        c[i] = div(a[i], b[i])
        d[i] = rem(a[i], b[i])
    end
end

function divrem3!(a, b, c, d)
    for i = 1:100_000
        c[i] = div(a[i], b[i])
        d[i] = a[i] - b[i] * c[i]
    end
end

a, b = rand(0:500_000, 100_000), rand(1:250_000, 100_000);
c, d = zeros(Int, 100_000), zeros(Int, 100_000)


julia> @benchmark divrem!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     834.000 μs (0.00% GC)
  median time:      837.300 μs (0.00% GC)
  mean time:        851.657 μs (0.00% GC)
  maximum time:     1.048 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem2!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.479 ms (0.00% GC)
  median time:      1.480 ms (0.00% GC)
  mean time:        1.499 ms (0.00% GC)
  maximum time:     1.716 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem3!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     839.100 μs (0.00% GC)
  median time:      843.400 μs (0.00% GC)
  mean time:        856.651 μs (0.00% GC)
  maximum time:     1.047 ms (0.00% GC)
  --------------
  samples:          100
  evals/sample:     1
```

### CODE REVIEW

**Comment by oscardssmith** [comment]
Have you tested speed for `BigInt`? I assume it will still be faster, but it often can cause corner cases. Other than that, this looks great!

**Comment by musm** [comment]
> Added misc changes like: Removed useless spaces in base\div.jl

These changes are orthogonal to the PR.

**Comment by r7rohan** [comment]
There was a trivial typo error fixed it.

Tested it for random BigInt, they show the same trend.
```
a, b = rand(0:500000000000000000000000000000000000000000000000000000000000000000000000000, 100000), rand(1:500000000000000000000000000000000000000000000000000000000000000000000000000, 100000);
c, d = zeros(BigInt, 100_000), zeros(BigInt, 100_000);

julia> @benchmark divrem!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  9.92 MiB
  allocs estimate:  500000
  --------------
  minimum time:     28.940 ms (0.00% GC)
  median time:      31.278 ms (0.00% GC)
  mean time:        46.815 ms (18.10% GC)
  maximum time:     435.204 ms (51.40% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem2!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  9.92 MiB
  allocs estimate:  500000
  --------------
  minimum time:     32.511 ms (0.00% GC)
  median time:      36.163 ms (0.00% GC)
  mean time:        50.150 ms (16.95% GC)
  maximum time:     418.931 ms (53.83% GC)
  --------------
  samples:          100
  evals/sample:     1

julia> @benchmark divrem3!($a,$b,$c,$d) samples=100
BenchmarkTools.Trial:
  memory estimate:  16.02 MiB
  allocs estimate:  649972
  --------------
  minimum time:     31.629 ms (0.00% GC)
  median time:      33.585 ms (0.00% GC)
  mean time:        61.267 ms (24.99% GC)
  maximum time:     448.482 ms (51.84% GC)
  --------------
  samples:          100
  evals/sample:     1
```

**Comment by Keno** [comment]
LGTM in the abstract, but CI is complaining about the tests.

File: test/int.jl [hunk 1]
```diff
@@ -339,8 +339,8 @@ let i = MyInt26779(1)
 end
 
 @testset "rounding division" begin
-    for x = -100:100
-        for y = 1:100
+    for x = -50:50
```
Comment (musm) [inline test/int.jl:342]: why change these existing tests

Comment (r7rohan) [reply test/int.jl:342]: Probably there is no need so reverting these changes

Comment (r7rohan) [reply test/int.jl:342]: Probably there is no need, so reverting this change

File: base/div.jl [hunk 2]
```diff
@@ -133,62 +132,76 @@ julia> divrem(7,3)
...
+        d = div(a, b, r)
+        (d, a - b*d)
     end
 end
 function divrem(x::Integer, y::Integer, rnd::typeof(RoundNearest))
     (q, r) = divrem(x, y)
     if x >= 0
         if y >= 0
-            r >=        (y÷2) + (isodd(y) | iseven(q)) ? (q+true, r-y) : (q, r)
+            r >=  (y÷2) + (isodd(y) | iseven(q)) ? (q+true, r-y) : (q, r)
```
Comment (musm) [inline base/div.jl:164]: As mentioned, these are unrelated style changes. The last person who wrote this  intended to line up `(y÷2)`. Since you're not modifying the code in these methods,  it's best that  these whitespace changes  be reverted.

**Comment by r7rohan** [comment]
Please merge

**Comment by oscardssmith** [comment]
@Keno good to merge now that CI is green?

File: base/div.jl [hunk 3]
```diff
@@ -133,15 +133,29 @@ julia> divrem(7,3)
 ```
 """
 divrem(x, y) = divrem(x, y, RoundToZero)
+
+function divrem(a::T, b::T, r::RoundingMode) where {T<:AbstractFloat}
```
Comment (JeffBezanson) [inline base/div.jl:137]:
```suggestion
function divrem(a::AbstractFloat, b::AbstractFloat, r::RoundingMode)
```

Otherwise the method below will be called for different-type floats, which I think we don't want?

Comment (r7rohan) [reply base/div.jl:137]: I wanted to write it for Float16, Float32, Float64, BigFloat - all the subtypes of AbstractFloat. I don't know what the difference will be

File: base/div.jl [hunk 4]
```diff
@@ -133,15 +133,30 @@ julia> divrem(7,3)
 ```
 """
 divrem(x, y) = divrem(x, y, RoundToZero)
+
+#avoids calling rem for Integers-Integers,
+#a-d*b not precise for Floats - AbstractFloat, AbstractIrrational. Rationals are still slower
+function divrem(a::Integer, b::Integer, r::RoundingMode)
```
Comment (Keno) [inline base/div.jl:139]: I think this should use explicit types on `r` for the two special cases, together with the note to remove them in 2.0.

**Comment by Keno** [comment]
This could use a PkgEval before merging. I know it doesn't seem like it, but the fallback chains for `div`-adjacent methods are very fragile because we allow packages to intercept both the 2 and 3-argument variants and want to provide various fallback. This needs to be cleaned up in 2.0, but not much we can do until then.

**Comment by r7rohan** [comment]
Sorry for late reply, I don't know much about PkgEval...so would appreciate help in that part. Apart from that, it seems to work well for divrem int-int types

**Comment by GregPlowman** [comment]
Does this just need PkgEval?

**Comment by oscardssmith** [comment]
Yeah. @Keno can you run it?

**Comment by simeonschaub** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-792378088):
```
NanosoldierError: failed to run tests: Build for julia on x86_64-linux-gnu did not complete successfully

```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_5fab42a/report.md)
cc @maleadt

**Comment by maleadt** [comment]
> curl: (6) Could not resolve host: github.com

Let's try again:

@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-792908248) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_86387a8/report.md). cc @maleadt

**Comment by KristofferC** [comment]
@nanosoldier `runtests(["AlphaStableDistributions", "BoltzmannMachines", "FilesystemDatastructures", "GPLinearODEMaker", "GeoDataFrames", "JetPackWaveFD", "Jusdl", "KissMCMC", "RiskAdjustedLinearizations", "Tar", "TensorToolbox"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/39045#issuecomment-793504165) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/24d688a_vs_faa3d41/report.md). cc @maleadt

File: base/div.jl [hunk 5]
```diff
@@ -157,6 +159,25 @@ function divrem(a, b, r::RoundingMode)
...
+    if r === RoundToZero
+        # For compat. Remove in 2.0.
+        d = div(a, b)
+        (d, a - d*b)
+    elseif r === RoundDown
+        # For compat. Remove in 2.0.
+        d = fld(a, b)
+        (d, a - d*b)
+    elseif r === RoundUp
+        # For compat. Remove in 2.0.
```
Comment (vtjnash) [inline base/div.jl:176]: What branch is not deprecated?

**Comment by vtjnash** [comment]
Seems to be working, but I'm confused by the line comments: why is all of the new code deprecated and marked for deletion in v2?

**Comment by r7rohan** [comment]
@vtjnash you are free to make the changes, I'm out of touch. AFAIK its working but Keno would be able to say better.

**Comment by vtjnash** [comment]
It seems to be working, I'm just confused by the content of the comments in the code, and was hoping you could clarify what they meant (I can make minor changes, if that's what's needed).

**Comment by StefanKarpinski** [comment]
Bump. Clarification requested...

**Comment by Keno** [comment]
In v2, all primitive integer types should implement three argument div(rem) and the two argument versions should just be fallbacks to those. At the moment, we have various complicated messes to handle situations where people overload the various two argument methods.

**Comment by oscardssmith** [comment]
Is this ready to be merged? If not, what does it need?
